query/src/sql/optimizer/heuristic/prune_columns.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+
+use crate::sql::optimizer::ColumnSet;
+use crate::sql::optimizer::SExpr;
+use crate::sql::plans::Aggregate;
+use crate::sql::plans::EvalScalar;
+use crate::sql::plans::LogicalGet;
+use crate::sql::plans::Project;
+use crate::sql::plans::RelOperator;
+use crate::sql::ScalarExpr;
+
+pub fn prune_columns(expr: &SExpr) -> Result<SExpr> {
+    match expr.plan() {
+        // For project and aggregate, collect required columns for its child
+        RelOperator::Project(p) => Ok(SExpr::create_unary(
+            RelOperator::Project(p.clone()),
+            keep_required_columns(expr.child(0)?, p.columns.clone())?,
+        )),
+        RelOperator::Aggregate(p) => {
+            let mut used = p.group_items.iter().fold(ColumnSet::new(), |acc, v| {
+                acc.union(&v.scalar.used_columns()).cloned().collect()
+            });
+            used = p.aggregate_functions.iter().fold(used, |acc, v| {
+                acc.union(&v.scalar.used_columns()).cloned().collect()
+            });
+            Ok(SExpr::create_unary(
+                RelOperator::Aggregate(p.clone()),
+                keep_required_columns(expr.child(0)?, used)?,
+            ))
+        }
+        // For the other plan nodes, keep searching for Project node.
+        p => {
+            let children = expr
+                .children()
+                .iter()
+                .map(prune_columns)
+                .collect::<Result<Vec<_>>>()?;
+            Ok(SExpr::create(p.clone(), children, None))
+        }
+    }
+}
+
+/// Keep columns referenced by parent plan node.
+/// `required` contains columns referenced by its ancestors. When a node has multiple children,
+/// the required columns for each child could be different and we may include columns not needed
+/// by a specific child. Columns should be skipped once we found it not exist in the subtree as we
+/// visit a plan node.
+fn keep_required_columns(expr: &SExpr, mut required: ColumnSet) -> Result<SExpr> {
```

Comment:
The initial required columns should be the columns used in result set.

For a query `select a, b from t`, the columns `a` and `b` are both required column. You can get the information from the root `BindContext` of a query plan.